### PR TITLE
Introduce ReferenceCell::get_default_linear_mapping() with triangulation argument. Use it.

### PR DESCRIPTION
**Pull request by bangerth**
A better approach than #11512. This also changes all uses of `StaticMappingQ1` in `source/` to the new function -- fixing a few potential bugs in the process, and leaving some TODOs that can't be addressed with the current finite element and quadrature interfaces. I'll convert `include/` later, but wanted to put this out here already for discussion.

/rebuild

### CODE REVIEW

File: source/grid/reference_cell.cc [hunk 1]
```diff
@@ -140,6 +140,25 @@ namespace ReferenceCell
 
 
 
+  template <int dim, int spacedim>
+  const Mapping<dim, spacedim> &
+  get_default_linear_mapping(const Triangulation<dim, spacedim> &triangulation)
+  {
+    const auto reference_cell_types = triangulation.get_reference_cell_types();
```
Comment (peterrum) [inline source/grid/reference_cell.cc:147]:
```suggestion
    const auto & reference_cell_types = triangulation.get_reference_cell_types();
```

**Review by peterrum — APPROVED** [review]
Looks good to me.

File: include/deal.II/fe/mapping_q1.h [hunk 2]
```diff
@@ -80,6 +80,10 @@ class MappingQ1 : public MappingQGeneric<dim, spacedim>
  * these contexts throughout the library, this class defines a static $Q_1$
  * mapping object. This object can then be used in all of those places where
  * such an object is needed.
+ *
+ * @note The use of this object should be avoided since it is only applicable
+ *   in cases where a mesh consists exclusively of quadrilaterals or hexahedra.
+ *   Use ReferenceCell::get_default_linear_mapping() instead.
```
Comment (tjhei) [inline include/deal.II/fe/mapping_q1.h:86]: Is this a candidate for DEPRECATED_EARLY? Or is this too early? :-)

Comment (bangerth) [reply include/deal.II/fe/mapping_q1.h:86]: Yes, somewhere in this series of patches.

**Review by tjhei — APPROVED** [review]

**Pull request merged**